gh-89610: Add syntax highlighting for `.pyi` files in IDLE

AlexWaygood opened the pull request:
This commit adds syntax highlighting for `.pyi` files in IDLE.
It also adds an option to open Python stub files in the "File-open" dialogue,
and an option to save a file as a `.pyi` file in the "File-save" dialogue.

<!-- gh-issue-number: gh-89610 -->
* Issue: gh-89610
<!-- /gh-issue-number -->

AlexWaygood committed:
bpo-45447: Add syntax highlighting for `.pyi` files in IDLE

This commit adds syntax highlighting for `.pyi` files in IDLE.
It also adds an option to open Python stub files in the "File-open" dialogue,
and an option to save a file as a `.pyi` file in the "File-save" dialogue.

E-Paine commented on Lib/idlelib/browser.py:
```diff
@@ -161,21 +162,19 @@ def GetSubList(self):
         dir, base = os.path.split(self.file)
         name, ext = os.path.splitext(base)
-        if os.path.normcase(ext) != ".py":
+        if not is_python_extension(os.path.normcase(ext)):
```
What's the reasoning for using `is_python_source` above and `is_python_extension` here?

AlexWaygood replied:
I think I was just trying to stick as closely to the existing source code as possible. `is_python_source`  will return `True` if the filepath points to a directory, and there's no logic like that in `listchildren` currently. Would it be better to use `is_python_source` instead of `is_python_extension` here, do you think?

AlexWaygood replied:
The reason I used `is_python_source` in `IsExpandable()` was because that was the guidance Terry gave me in the BPO issue.

E-Paine replied:
@terryjreedy I think it would make sense to use `is_python_extension` here. Your opinion?

E-Paine replied:
Ping @terryjreedy

terryjreedy replied:
ispythonsource is intended to include python files without .py? extension but starting with (re) '#!.*python'.  I need to review the usage of list children to see what I think its logic should be.  I may decide to leave moving .ispythonsource out of this issue (see issue), but have not done so yet.

terryjreedy replied:
If I cannot or do not decide soon, PR could be reduced to .pyi addition.

terryjreedy replied:
See comment for OnDoubleClick.  I will edit files and currently intend to leave no-ext files alone for now.

E-Paine commented on Lib/idlelib/util.py:
```diff
@@ -0,0 +1,36 @@
+PYTHON_EXTENSIONS = frozenset({'.py', '.pyw', '.pyi'})
+
+
+def is_python_extension(extension, is_valid=PYTHON_EXTENSIONS.__contains__):
```
Do we need the `is_valid` argument? I'd also prefer if we just passed the frozen set to that argument and the next line was just `extension in is_valid`, since I'm not a fan of accessing the dunder method in this way.

AlexWaygood replied:
It's not necessary, no -- I just thought it would make it slightly more performant to load the method at function-definition time. But I agree it's a little quirky; I'll take it out.

E-Paine commented on Lib/idlelib/idle_test/test_util.py:
```diff
@@ -0,0 +1,15 @@
+    def test_extensions(self):
+        self.assertIn('.pyi', util.PYTHON_EXTENSIONS)
+        self.assertIn('.py', util.PYTHON_EXTENSIONS)
+        self.assertIn('.pyw', util.PYTHON_EXTENSIONS)
```
It may make sense to test via the `is_python_extension` method instead.

AlexWaygood replied:
I've just rewritten the tests fairly extensively :)

E-Paine replied:
LGTM

E-Paine commented on Misc/NEWS.d/next/IDLE/2021-10-14-16-55-03.bpo-45447.FhiH5P.rst:
```diff
@@ -0,0 +1,2 @@
+IDLE now knows that `.pyi` files can only contain valid Python expressions,
+and will apply syntax highlighting accordingly. Patch by Alex Waygood.
```
```suggestion
IDLE now applies syntax highlighting to `.pyi` files. Patch by Alex Waygood.
```

E-Paine commented on Lib/idlelib/util.py:
```diff
@@ -0,0 +1,36 @@
+    * Python versions (editor and help_about),
+    * tk version and patchlevel (pyshell, help_about, maxos?, editor?),
+    * std streams (pyshell, run),
+    * warning stuff (ditto).
```
```suggestion
    * warning stuff (pyshell, run).
```
Make it explicit, so this line doesn't need to be changed when the point above is completed.

AlexWaygood committed:
Improve news entry.

Co-authored-by: E-Paine <63801254+E-Paine@users.noreply.github.com>

AlexWaygood committed:
Improve docstring at top of util.py

Co-authored-by: E-Paine <63801254+E-Paine@users.noreply.github.com>

AlexWaygood committed:
Improve tests

AlexWaygood committed:
Use keyword arguments in test_util.py

AlexWaygood committed:
Correct alphabetical order in README.txt...

AlexWaygood committed:
Update coverage statistics in test files

AlexWaygood committed:
Add __main__ block for util.py, tweak added test for iomenu.py, tweak docstrings to keep line length within 72 chars

AlexWaygood committed:
Tweak test_util.py

AlexWaygood committed:
Merge branch 'python:main' into idle-pyi-syntax-highlighting

terryjreedy commented on Lib/idlelib/util.py:
```diff
@@ -0,0 +1,40 @@
+from os import path
+
+# .pyw is for Windows; .pyi is for stub files.
+PYTHON_EXTENSIONS = frozenset({'.py', '.pyw', '.pyi'})
```
```suggestion
PYTHON_EXTENSIONS = frozenset(('.py', '.pyw', '.pyi'))
```
For why, consider
```
>>> dis('frozenset((1,2,3))')
  1           0 LOAD_NAME                0 (frozenset)
              2 LOAD_CONST               0 ((1, 2, 3))
              4 CALL_FUNCTION            1
              6 RETURN_VALUE
>>> dis('frozenset({1,2,3})')
  1           0 LOAD_NAME                0 (frozenset)
              2 BUILD_SET                0
              4 LOAD_CONST               0 (frozenset({1, 2, 3}))
              6 SET_UPDATE               1
              8 CALL_FUNCTION            1
             10 RETURN_VALUE
>>> dis('{1,2,3}')
```

terryjreedy commented on Lib/idlelib/idle_test/test_browser.py:
```diff
@@ -1,4 +1,4 @@
-"Test browser, coverage 90%."
+"Test browser, coverage 81%."
```
I don't believe this.  IDLE coverage requires a customized .coveragerc file, as discussed either in README or idle_test/README (and in an open issue that should be finished).  I will recheck all these numbers.

AlexWaygood replied:
OK. Would you like me to take out those changes for now?

terryjreedy commented on Lib/idlelib/idle_test/test_util.py:
```diff
@@ -0,0 +1,98 @@
+from idlelib import util
+
+
+class ExtensionTest(unittest.TestCase):
```
I will condense the extension tests.

terryjreedy reviewed (changes requested):
Besides inline comments, add file requested on issue.

AlexWaygood committed:
Update Lib/idlelib/util.py

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

AlexWaygood committed:
Add example `.pyi` file

AlexWaygood committed:
Revert changes to coverage statistics

AlexWaygood committed:
Rename `example.pyi` to `example_stub.pyi`

AlexWaygood committed:
Have stub files combined with other Python files in `iomenu.py`

AlexWaygood committed:
Merge remote-tracking branch 'origin/main' into idle-pyi-syntax-highlighting

AlexWaygood committed:
Fix broken test, improve other tests

AlexWaygood committed:
Add `example_noext` file

AlexWaygood committed:
Add entry to 'What's new in 3.11'

terryjreedy committed:
Merge remote-tracking branch 'upstream/main' into pr_28950

terryjreedy commented on Lib/idlelib/browser.py:
```diff
@@ -161,21 +162,19 @@ def GetSubList(self):
-            return
-        file_open(self.file)
+        if self.IsExpandable() and os.path.exists(self.file):
+            file_open(self.file)
```
For no-extension files, IsExpandable requires that the file be open.  It make no sense to test that a file exists after is has been opened.  Nor to close and reopen after the uneeded test.  Testing no-extension files should be separate.

terryjreedy reviewed (commented):
Alex, I am editing the changes, so leave alone for now.

terryjreedy committed:
Merge remote-tracking branch 'upstream/main' into pr_28950

terryjreedy committed:
Revert changes in browser.py

terryjreedy committed:
Revert changes to editor.EditorWindow.ispythonsource.

terryjreedy committed:
Revise util, removing new functions.

terryjreedy committed:
Revise test_util, deleting most tests.

terryjreedy committed:
Use tuple instead of frozen set.

This enables computing text for open/save dialogs.
.py must be first to make it default on non-macOS.

terryjreedy committed:
Update 2021-10-14-16-55-03.bpo-45447.FhiH5P.rst

terryjreedy committed:
Create 3.11.rst

terryjreedy committed:
Update 3.11.rst

terryjreedy committed:
Revert unwanted changed to what's new

AlexWaygood committed:
Remove commented-out line

terryjreedy committed:
Merge remote-tracking branch 'upstream/main' into pr_28950

terryjreedy committed:
Final tweaks.

terryjreedy committed:
Merge branch 'idle-pyi-syntax-highlighting' of https://github.com/AlexWaygood/cpython into pr_28950
